File: src/com/example/javas/medium/M003_ParkingLotSystem.java
/*
SYSTEM DESIGN PROBLEM:
Design a simple Parking Lot System (Simplified for 25 min interview)

REQUIREMENTS:
- Park a vehicle (assign next available spot)
- Remove a vehicle from a parking spot
- Check available spots

CONSTRAINTS:
- Fixed capacity (e.g., 10 spots)
- Simple sequential spot assignment (no need for min-heap)
- Use HashMap for O(1) operations

DESIGN APPROACH:
1. Use HashMap to store vehicle ID → spot number
2. Use simple counter for next available spot
3. Track occupied spots count

DATA STRUCTURES USED:
- HashMap<String, Integer>: vehicleId → spotNumber
- int nextSpot: next available spot to assign

TIME COMPLEXITY:
- park(): O(1)
- remove(): O(1)
- getAvailableSpots(): O(1)

SPACE COMPLEXITY: O(n) where n is capacity
*/

package com.example.javas.medium;

import java.util.*;

public class M003_ParkingLotSystem {
    
    // Simplified Parking Lot System
    static class ParkingLot {
        private int capacity;
        private int occupiedSpots;
        private HashMap<String, Integer> vehicleToSpot; // vehicleId → spot number
        
        public ParkingLot(int capacity) {
            // DONE: Initialize capacity, occupiedSpots (0), and HashMap
            this.capacity = capacity;
            this.occupiedSpots = 0;
            this.vehicleToSpot = new HashMap<>();
        }
        
        public int park(String vehicleId) {
            // DONE: Implement park operation
            // 1. Check if parking lot is full (occupiedSpots >= capacity)
            // 2. Check if vehicle already parked (vehicleId exists in map)
            // 3. Assign spot (occupiedSpots + 1), add to map, increment occupiedSpots
            // 4. Return spot number (or -1 if full/duplicate)
            
            if (occupiedSpots >= capacity) {
                return -1; // Parking lot is full
            }
            
            if (vehicleToSpot.containsKey(vehicleId)) {
                return -1; // Vehicle already parked
            }
            
            int spotNumber = occupiedSpots + 1;
            vehicleToSpot.put(vehicleId, spotNumber);
            occupiedSpots++;
            
            return spotNumber;
        }
        
        public boolean remove(String vehicleId) {
            // DONE: Implement remove operation
            // 1. Check if vehicle exists in map
            // 2. If yes: remove from map, decrement occupiedSpots, return true
            // 3. If no: return false
            
            if (!vehicleToSpot.containsKey(vehicleId)) {
                return false; // Vehicle not found
            }
            
            vehicleToSpot.remove(vehicleId);
            occupiedSpots--;
            
            return true;
        }
        
        public int getAvailableSpots() {
            // DONE: Return capacity - occupiedSpots
            return capacity - occupiedSpots;
        }
        
        public void displayStatus() {
            System.out.println("Occupied: " + occupiedSpots + "/" + capacity);
            System.out.println("Available: " + getAvailableSpots());
            System.out.println("Vehicles: " + vehicleToSpot);
        }
    }
    
    // TEST CASES
    public static void main(String[] args) {
        ParkingLot lot = new ParkingLot(3);
        
        System.out.println("Park CAR-1: " + lot.park("CAR-1"));
        System.out.println("Park CAR-2: " + lot.park("CAR-2"));
        System.out.println("Park CAR-3: " + lot.park("CAR-3"));
        System.out.println("Park CAR-4 (full): " + lot.park("CAR-4")); // Should be -1
        System.out.println("Available: " + lot.getAvailableSpots());
        System.out.println();
        
        System.out.println("Remove CAR-2: " + lot.remove("CAR-2"));
        System.out.println("Available: " + lot.getAvailableSpots());
        System.out.println("Remove CAR-999: " + lot.remove("CAR-999")); // Should be false
    }
}

File: pythonX/medium/m003_parking_lot_system.py
"""
M003: Design Parking Lot System (Simplified)

Summary:
Design a simple parking lot system for managing vehicle parking.

Requirements:
Implement the ParkingLot class:
- park(vehicle_id): Assigns next available spot
- remove(vehicle_id): Removes vehicle from parking lot
- get_available_spots(): Returns count of available spots

Approach:
Use dict to map vehicle_id -> spot number, track occupied count.

Data Structures:
- dict[str, int]: vehicle_id -> spot_number
- int: occupied_spots counter

Big O Analysis:
Time Complexity: O(1) for all operations
Space Complexity: O(n) where n is capacity
"""


class ParkingLot:
    
    def __init__(self, capacity: int):
        """
        TODO: Initialize parking lot
        Set capacity, occupied_spots = 0, create dict
        """
        pass
    
    def park(self, vehicle_id: str) -> int:
        """
        TODO: Park a vehicle
        1. Check if lot is full (occupied >= capacity)
        2. Check if vehicle already parked
        3. Assign spot (occupied + 1), add to dict, increment occupied
        4. Return spot number or -1 if full/duplicate
        """
        pass
    
    def remove(self, vehicle_id: str) -> bool:
        """
        TODO: Remove vehicle
        1. Check if vehicle exists
        2. Remove from dict, decrement occupied
        3. Return True if removed, False otherwise
        """
        pass
    
    def get_available_spots(self) -> int:
        """
        TODO: Get available spots
        Return capacity - occupied_spots
        """
        pass
    
    def display_status(self):
        """Display current status"""
        print(f"Occupied: {getattr(self, 'occupied_spots', 0)}/{getattr(self, 'capacity', 0)}")
        print(f"Available: {self.get_available_spots()}")
        print(f"Vehicles: {getattr(self, 'vehicle_to_spot', {})}")


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Basic parking and removal")
    lot = ParkingLot(3)
    print(f"Park CAR-1: {lot.park('CAR-1')}")       # Expected: 1
    print(f"Park CAR-2: {lot.park('CAR-2')}")       # Expected: 2
    print(f"Park CAR-3: {lot.park('CAR-3')}")       # Expected: 3
    print(f"Park CAR-4 (full): {lot.park('CAR-4')}")  # Expected: -1
    print(f"Available: {lot.get_available_spots()}")  # Expected: 0
    
    print("\nTest 2: Remove and park again")
    print(f"Remove CAR-2: {lot.remove('CAR-2')}")   # Expected: True
    print(f"Available: {lot.get_available_spots()}")  # Expected: 1
    print(f"Remove CAR-999: {lot.remove('CAR-999')}")  # Expected: False
